src/sentry/services/hybrid_cloud/organization/impl.py

```diff
@@ -512,6 +515,69 @@ def delete_option(self, *, organization_id: int, key: str) -> None:
+                organization = Organization.objects.get(id=auth_provider.organization_id)
+                # Deletes do not cascade immediately -- if we removed and add a new provider
+                # we should just clear that older provider.
+                AuthProviderReplica.objects.filter(organization=organization).exclude(
+                    auth_provider_id=auth_provider.id
+                ).delete()
+                existing = AuthProviderReplica.objects.filter(
+                    auth_provider_id=auth_provider.id
+                ).first()
+                update = {
+                    "organization": organization,
+                    "provider": auth_provider.provider,
+                    "config": auth_provider.config,
+                    "default_role": auth_provider.default_role,
+                    "default_global_access": auth_provider.default_global_access,
+                    "allow_unlinked": auth_provider.flags.allow_unlinked,
+                    "scim_enabled": auth_provider.flags.scim_enabled,
+                }
+
+                if not existing:
+                    AuthProviderReplica.objects.create(auth_provider_id=auth_provider.id, **update)
+                    return
+
+                existing.update(**update)
+        except Organization.DoesNotExist:
+            return
+
+    def upsert_replicated_auth_identity(
+        self, *, auth_identity: RpcAuthIdentity, region_name: str
+    ) -> None:
+        with enforce_constraints(transaction.atomic(router.db_for_write(AuthIdentityReplica))):
+            # Since coalesced outboxes won't replicate the precise ordering of changes, these
+            # unique keys can cause a deadlock in updates.  To address this, we just delete
+            # any conflicting items and allow future outboxes to carry the updates
+            # for the auth identities that should follow (given they will share the same shard).
+            AuthIdentityReplica.objects.filter(
+                ident=auth_identity.ident,
+                auth_provider_id=auth_identity.auth_provider_id,
+            ).exclude(auth_identity_id=auth_identity.id).delete()
+            AuthIdentityReplica.objects.filter(
+                user_id=auth_identity.user_id,
+                auth_provider_id=auth_identity.auth_provider_id,
+            ).exclude(auth_identity_id=auth_identity.id).delete()
+
+            existing = AuthIdentityReplica.objects.filter(auth_identity_id=auth_identity.id).first()
+            update = {
+                "user_id": auth_identity.user_id,
+                "auth_provider_id": auth_identity.auth_provider_id,
+                "ident": auth_identity.ident,
+                "data": auth_identity.data,
+            }
+
+            if not existing:
+                AuthIdentityReplica.objects.create(auth_identity_id=auth_identity.id, **update)
+                return
+
+            existing.update(**update)
```

Comment:
You could use Django's `AuthIdentityReplica.objects.create_or_update` method to do the upsert operation as well.

Reply:
You may be right, I'll revisit.